File: python/day10.py
from aoc import *


def find_trailheads(grid):
    return {k for k, v in grid.items() if v == 0}

def dfs(grid, start):
    ends = []
    stack = [(start, 0)]
    while stack:
        curr, h = stack.pop()
        if h == 9:
            ends.append(curr)
            continue
        nbs = [(pt, pt_h) for pt in neighbours(*curr)
               if (pt_h := grid.get(pt)) == h + 1]
        stack.extend(nbs)
    return len(set(ends)), len(ends)

def calc_score(scores, idx):
    return sum(score[idx] for score in scores)

def solve(filename):
    grid = list2grid(parse_lines(read_input(filename), digits))
    trailheads = find_trailheads(grid)
    scores = [dfs(grid, start) for start in trailheads]
    return calc_score(scores, 0), calc_score(scores, 1)


solve(10)

File: elixir/day10.exs
defmodule Day10 do
  defp parse_input(input) do
    import AOC

    input
    |> parse_lines(:digits)
    |> grid_to_point_map
  end

  defp find_trailheads(grid) do
    import Map

    grid
    |> filter(fn {_, v} -> v == 0 end)
    |> keys
  end

  defp dfs_aux(_grid, [], ends) do
    [ends |> Enum.uniq() |> length, ends |> length]
  end

  defp dfs_aux(grid, [curr | rest], ends) do
    curr_h = Map.get(grid, curr)

    if curr_h == 9 do
      dfs_aux(grid, rest, [curr | ends])
    else
      nbs = AOC.neighbours_4(curr, &(Map.get(grid, &1, 0) == curr_h + 1))
      dfs_aux(grid, nbs ++ rest, ends)
    end
  end

  defp dfs(grid, start) do
    dfs_aux(grid, [start], [])
  end

  def solve(filename) do
    import Enum

    grid =
      filename
      |> AOC.read_input()
      |> parse_input

    grid
    |> find_trailheads
    |> map(&dfs(grid, &1))
    |> zip
    |> map(&Tuple.sum/1)
  end
end

Day10.solve(10)
